### PR TITLE
Replace plugin.args_invariant_decorator_callback with use of ParamSpec

### LINKED ISSUE #90: trio_typing.plugin incompatible with mypy 1.7's new type inference
When the `trio_typing.plugin` is enabled with mypy 1.7, it can cause false positive errors (even when trio is not even in use) related to Mypy's new type inference which was enabled by default in 
https://github.com/python/mypy/pull/16345 for [mypy 1.7](https://mypy-lang.blogspot.com/2023/11/mypy-17-released.html). 

When linting async functions that take generic callables, the new type inference appears to automatically infer the variance of typevars. For example:
https://mypy-play.net/?mypy=latest&python=3.11&gist=73fc0a0b94a3f2410080c0482df7ecab
Outputs:
```
main.py:24: note: Revealed type is "def () -> __main__.C"
main.py:25: note: Revealed type is "def [_T_co] (make_client: def () -> _T_co`173) -> contextlib._AsyncGeneratorContextManager[_T_co`173]"
main.py:26: note: Revealed type is "contextlib._AsyncGeneratorContextManager[__main__.C]"
```

But if you add `# mypy: old-type-inference` to the top: https://mypy-play.net/?mypy=latest&python=3.11&gist=be6764bbead8f238976dd1be12d8d7e8 then it outputs:

```
main.py:25: note: Revealed type is "def () -> __main__.C"
main.py:26: note: Revealed type is "def [T] (make_client: def () -> T`-1) -> contextlib._AsyncGeneratorContextManager[T`-1]"
main.py:27: note: Revealed type is "contextlib._AsyncGeneratorContextManager[__main__.C]"
```

When the _new type inference_ is used with a config file that enables the plugin
```
# mypy.ini
[mypy]
plugins = trio_typing.plugin
```

then an error is reported for this code:
```python
    client: C
    async with _do_open_client(make_client) as client:
```
```
example.py:27: error: Incompatible types in assignment (expression has type "_T_co", variable has type "C")  [assignment]
```

### PR DESCRIPTION
**Pull request by oremanj**
Resolves #90, a regression that appeared with mypy 1.7. In order to have ParamSpec available, we now require at least mypy 1.0. `args_invariant_decorator_callback` is modeled off the mypy `contextlib` plugin which was replaced with use of ParamSpec in Oct 2022 (https://github.com/python/mypy/commit/3108669b0c1d57a6eecf84002fa1cf240449dba6); mypy 1.0 came out in Feb 2023.

The test described in #90 no longer fails now that this hook has been removed.

While we're here, update stubs to be consistent with Trio's new inline stubs to satisfy CI. This included:
* giving names to the statistics types
* making some SocketType attributes into properties
* listing out the arguments of `__aexit__` instead of using a star-args shortcut
* fixing the signature of SocketStream.setsockopt

### CODE REVIEW

**Review by CoolCat467 — APPROVED** [review]
This looks good on a surface level, but I haven't checked that these match with the real project.

**Comment by oremanj** [comment]
I solemnly swear that all of the statistics classes were copy-pasted from current trio. :-) stubtest is happy so I think there's relatively little that could be incorrect.

**Comment by CoolCat467** [comment]
100% approval then!

**Review by A5rocks — APPROVED** [review]
Beyond these issues, looks fine.

(but I may have missed something)

File: trio-stubs/lowlevel.pyi [hunk 1]
```diff
@@ -66,11 +68,42 @@ class UnboundedQueue(Generic[_T], metaclass=ABCMeta):
...
+        tasks_waiting_read: int = attr.ib()
+        tasks_waiting_write: int = attr.ib()
+        backend: Literal["epoll"] = attr.ib(init=False, default="epoll")
+
+else:  # kqueue
+    @attr.frozen
+    class _IOStatistics:
+        tasks_waiting: int = attr.ib()
+        monitors: int = attr.ib()
+        backend: Literal["kqueue"] = attr.ib(init=False, default="kqueue")
```
[the remark this answers is not in the document]
Comment (oremanj) [reply trio-stubs/lowlevel.pyi:97]: init= impacts the signature of the generated `__init__` which stubtest will notice mismatches in

**Review by CoolCat467 — APPROVED** [review]

**Pull request merged**